File: Graph(그래프)/DFS(깊이 우선 탐색)/백준_2331_반복수열/백준_2331_반복수열.java
import java.util.*;

class Main {
  static int[] count_arr = new int[300000];

  static void DFS(int a, int p){
    double new_a = 0;
    while( a> 0){
      double left = a %10;
      new_a += Math.pow(left, p);
      a /=10;
    }

    int x = (int) new_a;
    count_arr[x]++;
    if(count_arr[x] > 2) return;
    DFS(x,p);
  
  }
  
  public static void main(String[] args) {
    Scanner s = new Scanner(System.in);
    int A = s.nextInt(); 
    int P = s.nextInt();

    count_arr[A]++;
    DFS(A, P);
    int count = 0;
    for(int i =0 ; i< count_arr.length; i++)
      if(count_arr[i] == 1) count++;

    System.out.println(count);
    

    
  }
}

File: Graph(그래프)/DFS(깊이 우선 탐색)/백준_2331_반복수열/백준_2331_반복수열.kt
import java.util.*

var count_arr = Array<Int>(300000) {0}

fun DFS(a: Int, p: Int){        //주의! 코를린 함수의 매개변수는 무조건 val!
    var x = a                   //a값 변경을 위해 var변수 생성
    var new_a = 0
    while(x > 0){
        var left = x % 10
        var leftsquare = 1
        for(i in 1..p) leftsquare *= left       //pow함수 쓰지 않고 for문으로 직접 제곱수 구함
        new_a += leftsquare
        x /= 10
    }

    count_arr[new_a]++                          //새로운 수. 배열에 체크
    if(count_arr[new_a] > 2) return
    DFS(new_a, p)
}

fun main(args: Array<String>) {         //var - 변경가능, val - 변경불가 변수
    val s = Scanner(System.`in`)
    val A = s.nextInt(); val P = s.nextInt()

    count_arr[A]++
    DFS(A, P);
    var count = 0
    for(i in count_arr)
        if(i == 1) count++

    print("${count}")


}